pkg/pipeline/stream/source_test.go

```diff
@@ -0,0 +1,174 @@
+// Copyright © 2022 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package stream
+
+import (
+	"context"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/conduitio/conduit/pkg/connector/mock"
+	"github.com/conduitio/conduit/pkg/foundation/metrics/noop"
+	"github.com/conduitio/conduit/pkg/plugin"
+	"github.com/conduitio/conduit/pkg/record"
+	"github.com/golang/mock/gomock"
+	"github.com/google/uuid"
+	"github.com/matryer/is"
+)
+
+func TestSourceNode_Run(t *testing.T) {
+	is := is.New(t)
+	ctx := context.Background()
+	ctrl := gomock.NewController(t)
+
+	src, wantRecords := newMockSource(ctrl, 10, nil)
+	node := &SourceNode{
+		Name:          "source-node",
+		Source:        src,
+		PipelineTimer: noop.Timer{},
+	}
+	out := node.Pub()
+
+	nodeDone := make(chan struct{})
+	go func() {
+		defer close(nodeDone)
+		err := node.Run(ctx)
+		is.NoErr(err)
+	}()
+
+	for _, wantRecord := range wantRecords {
+		gotMsg := <-out
+		is.Equal(gotMsg.Record, wantRecord)
+		err := gotMsg.Ack()
+		is.NoErr(err)
+	}
+
+	// we stop the node now, but the mock will simulate as if it still needs to
+	// produce the records, the source should drain them
+	err := node.Stop(ctx, nil)
+	is.NoErr(err)
+
+	select {
+	case <-time.After(time.Second):
+		is.Fail() // expected node to close outgoing channel
+	case _, ok := <-out:
+		is.True(!ok) // channel out should be closed
+	}
+
+	select {
+	case <-time.After(time.Second):
+		is.Fail() // expected node to stop running
+	case <-nodeDone:
+		// all good
+	}
+}
+
+func TestSourceNode_StopWhileNextNodeIsStuck(t *testing.T) {
+	// A pipeline can't be stopped gracefully if the next node after a source
+	// node blocks forever (or for a long time), because we can't inject the
+	// stop message into the stream.
+	t.Skip("bug isn't fixed yet")
```

Comment:
IIUC, this isn't something we want to tackle in this PR. If that is the case, do we have an issue for this?